In [14]:
import numpy as np
import pandas as pd


In [15]:
# Dataset
from sklearn.datasets import fetch_california_housing

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [16]:
#Regression Models
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor

In [17]:
# Evaluation Metrics
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [18]:
# Supress warnings
import warnings
warnings.filterwarnings("ignore")
print("All libraries imorted successfully")


All libraries imorted successfully


# 1. Loading and Preprocessing

In [19]:
# Load the dataset from sklearn
housing = fetch_california_housing()
# Convert to pandas DataFrame
df = pd.DataFrame(housing.data, columns= housing.feature_names)
df['MedHouseVal'] = housing.target #Target coulumn : Median House Value
print("Shape:", df.shape)
df.head()

Shape: (20640, 9)


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


**1.2 Dataset Overview**

In [20]:
# General info
df.info()
# Statistical summary
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MedInc       20640 non-null  float64
 1   HouseAge     20640 non-null  float64
 2   AveRooms     20640 non-null  float64
 3   AveBedrms    20640 non-null  float64
 4   Population   20640 non-null  float64
 5   AveOccup     20640 non-null  float64
 6   Latitude     20640 non-null  float64
 7   Longitude    20640 non-null  float64
 8   MedHouseVal  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
count,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,3.870671,28.639486,5.429000,1.096675,1425.476744,3.070655,35.631861,-119.569704,2.068558
std,1.899822,12.585558,2.474173,0.473911,1132.462122,10.386050,2.135952,2.003532,1.153956
min,0.499900,1.000000,0.846154,0.333333,3.000000,0.692308,32.540000,-124.350000,0.149990
25%,2.563400,18.000000,4.440716,1.006079,787.000000,2.429741,33.930000,-121.800000,1.196000
50%,3.534800,29.000000,5.229129,1.048780,1166.000000,2.818116,34.260000,-118.490000,1.797000
75%,4.743250,37.000000,6.052381,1.099526,1725.000000,3.282261,37.710000,-118.010000,2.647250
max,15.000100,52.000000,141.909091,34.066667,35682.000000,1243.333333,41.950000,-114.310000,5.000010


* The describe output clearly shows features operate on vastly different scales - Population has a mean of 1,425 while AveBedrms has a mean of 1.09.
* Without StandardScaler, large-scale features would dominate and bias Linear Regression coefficients, leading to inaccurate predictions.

**1.3 Check for Missing Values**

In [21]:
df.isnull().sum()

MedInc         0
HouseAge       0
AveRooms       0
AveBedrms      0
Population     0
AveOccup       0
Latitude       0
Longitude      0
MedHouseVal    0
dtype: int64

* The dataset has no missing values. Therefore no imputation strategy was required.

**1.4 Feature Scaling (Standardization)**

In [22]:
X = df.drop('MedHouseVal', axis=1)   # Features
y = df['MedHouseVal']                 # Target

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled  = scaler.transform(X_test)

print("Train size:", X_train_scaled.shape)
print("Test size :", X_test_scaled.shape)

Train size: (16512, 8)
Test size : (4128, 8)


* As observed from df.describe(), features operate on vastly different scales — Population reaches up to 35,682 while AveBedrms stays between 0.33 and 1.09.
* Without StandardScaler, large-scale features would dominate and bias Linear Regression coefficients, leading to inaccurate predictions.
* StandardScaler transforms all features to mean = 0 and standard deviation = 1, ensuring every feature contributes equally during training.
* The dataset is split into 80% training (16,512 samples) and 20% testing (4,128 samples) using train_test_split. The scaler is fitted only on training data and applied to test data to prevent data leakage.

# 2. Regression Algorithm Implementation

**2.1 Linear Regression**

In [23]:
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)
y_pred_lr = lr.predict(X_test_scaled)
print("Linear Regression - Done")

Linear Regression - Done


* Linear Regression models the relationship between a dependent variable and one or more independent variables using a linear equation. It fits a hyperplane by minimizing the sum of squared errors.
* Formula: y = m₁x₁ + m₂x₂ + ... + mₙxₙ + c
* Linear Regression is suitable for California house price prediction because it can model the relationship between multiple housing features (such as income, rooms, and population) and price using a linear equation. MedInc has a strong correlation with house prices, making it a reliable and fast baseline model for this dataset.

**2.2 Decision Tree Regressor**

In [24]:
dt = DecisionTreeRegressor(random_state=42)
dt.fit(X_train_scaled, y_train)
y_pred_dt = dt.predict(X_test_scaled)

print("Decision Tree - Done")

Decision Tree - Done


* A Decision Tree splits the data into smaller groups by asking yes/no questions on feature values repeatedly until it reaches a final group, where it predicts the mean house price of that group. Each split is chosen to minimize MSE.
* Decision Tree is suitable for California house price prediction because it can capture non-linear relationships between features and prices. Unlike Linear Regression, it handles cases where house prices don't increase proportionally.

# 3. Model Evaluation and Comparison

In [25]:
def evaluate_model(name, y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    mae = mean_absolute_error(y_true, y_pred)
    r2 = r2_score(y_true, y_pred)
    return{"Model": name, "MSE": round(mse, 4),
           "MAE": round(mae, 4), "R2": round(r2, 4)}

results = [
    evaluate_model("Linear Regression", y_test, y_pred_lr),
    evaluate_model("Decision Tree",     y_test, y_pred_dt),
]
results_df = pd.DataFrame(results).set_index("Model")
print(results_df)

                      MSE     MAE      R2
Model                                    
Linear Regression  0.5559  0.5332  0.5758
Decision Tree      0.4940  0.4539  0.6230


* Decision Tree is the best performing model with the lowest MSE (0.4940), lowest MAE (0.4539) and highest R² (0.6230), as it captures the non-linear relationships between housing features and prices.
* Linear Regression is the worst performing model with the highest MSE (0.5559), highest MAE (0.5332) and lowest R² (0.5758). It is limited by its assumption of linearity and fails to capture the complex non-linear patterns in California housing data, resulting in less accurate predictions.